modules/cms/controllers/Index.php

```diff
@@ -397,10 +389,133 @@ public function onExpandMarkupToken()
+            // Populate the filesystem with the template and then remove it from the db
+            $datasource = $this->getThemeDatasource();
+            $datasource->pushToSource($template, 'filesystem');
+            $datasource->removeFromSource($template, 'database');
+
+            Flash::success(Lang::get('cms::lang.editor.commit_success', ['type' => $type]));
+        }
+
+        return array_merge($this->getUpdateResponse($template, $type), ['forceReload' => true]);
+    }
+
+    /**
+     * Resets a template to the version on the filesystem
+     *
+     * @return array $response
+     */
+    public function onReset()
+    {
+        $this->validateRequestTheme();
+        $type = Request::input('templateType');
+        $template = $this->loadTemplate($type, trim(Request::input('templatePath')));
+
+        if ($this->canResetTemplate($template)) {
+            // Remove the template from the DB
+            $datasource = $this->getThemeDatasource();
+            $datasource->removeFromSource($template, 'database');
+
+            Flash::success(Lang::get('cms::lang.editor.reset_success', ['type' => $type]));
+        }
+
+        return array_merge($this->getUpdateResponse($template, $type), ['forceReload' => true]);
+    }
+
     //
-    // Methods for the internal use
+    // Methods for internal use
     //
 
+    /**
+     * Get the response to return in an AJAX request that updates a template
+     *
+     * @param CmsObject $template The template that has been affected
+     * @param string $type The type of template being affected
+     * @return array $result;
+     */
+    protected function getUpdateResponse(CmsObject $template, string $type)
+    {
+        $result = [
+            'templatePath'  => $template->fileName,
+            'templateMtime' => $template->mtime,
+            'tabTitle'      => $this->getTabTitle($type, $template)
+        ];
+
+        if ($type === 'page') {
+            $result['pageUrl'] = Url::to($template->url);
+            $router = new Router($this->theme);
+            $router->clearCache();
+            CmsCompoundObject::clearCache($this->theme);
+        }
+
+        $result['canCommit'] = $this->canCommitTemplate($template);
+        $result['canReset'] = $this->canResetTemplate($template);
+
+        return $result;
+    }
+
+    /**
+     * Get the active theme's datasource
+     *
+     * @return \October\Rain\Halcyon\Datasource\DatasourceInterface
+     */
+    protected function getThemeDatasource()
+    {
+        return $this->theme->getDatasource();
+    }
+
+    /**
+     * Check to see if the provided template can be committed
+     * Only available in debug mode, the DB layer must be enabled, and the template must exist in the database
+     *
+     * @param CmsObject $template
+     * @return boolean
+     */
+    protected function canCommitTemplate(CmsObject $template)
+    {
+        $result = false;
+
+        if (Config::get('app.debug', false) &&
```

Comment:
Should the debug check be removed? It's already being checked in the `Theme::databaseLayerEnabled()` call.
```suggestion
        if (
```

Reply:
No, the check in `Theme::databaseLayerEnabled()` is only for when the config value is `null`. This check is here on purpose to prevent DB changes to be committed to the filesystem in a production environment as the committing process removes the DB copy which would cause a multi-app-server one-db-server setup to get out of sync as the content would now be updated on one instance's filesystem and removed from the other instance's datasource (the main database server).

Reply:
Ah yep, that makes sense.